### PR TITLE
[eas-build-job][build-tools][steps] Accept legacy command/path custom function shape in function.yml; rename references in the consumers

### PR DESCRIPTION
**Pull request by sswrk**
# Why

We want to grandfather the legacy reusable functions from custom builds (https://docs.expo.dev/custom-builds/functions/) into the new reusable functions that can be used in workflows. This is to make the migration from custom builds to workflows easier.

Currently, we only support composite functions. The first step towards extending the support is modifying the `eas-build-job` Zod schema.

# How

- Added the `legacyFunction` module with Zod schemas to match the legacy custom builds functions.
  - Two separate schemas: `LegacyCommandFunctionConfigZ` (functions with `command:` present) and `LegacyPathFunctionConfigZ` (functions with `path:` present).
- Extracted `compositeFunction` and created a new `localFunction` module with schema that is a union of composite function schema and legacy function schemas.
- Renamed references of "composite local functions" schema in `steps` and `build-tools` that should reference "local  functions".

# Test Plan

Added unit tests.

### CODE REVIEW

File: packages/eas-build-job/src/legacyFunction.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    .object({
+      name: z.string(),
+      required: z.boolean().optional(),
+    })
+    .strict(),
+]);
+
+const LegacyFunctionPlatformZ = z.enum(['darwin', 'linux']);
+
+const LegacyFunctionBaseZ = z.object({
```
Comment (hSATAC) [inline packages/eas-build-job/src/legacyFunction.ts:57]: Non-blocking, and this may be intentional: should the legacy schema also accept the camelCase spellings?

From what I can tell, those are the native keys in the old Joi schema and snake_case is added as a rename alias, so `supportedRuntimePlatforms`, `defaultValue`, `allowedValues`, and `allowedValueType` work in custom builds today. Here they fail with the generic union error:

```yaml
command: echo hi
supportedRuntimePlatforms: [darwin]
```

Maybe only grandfathering the documented snake_case shape is what we want. If so, could we add a rejection test to make that boundary explicit? Otherwise I think these branches should accept both.

File: packages/eas-build-job/src/localFunction.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+export type LocalFunctionConfig = CompositeFunctionConfig | LegacyFunctionConfig;
+
+export function isLegacyFunctionConfig(
+  config: LocalFunctionConfig
+): config is LegacyFunctionConfig {
+  return config.runs === undefined;
+}
+
+export type LocalFunctionCatalog = Record<string, LocalFunctionConfig>;
```
Comment (hSATAC) [inline packages/eas-build-job/src/localFunction.ts:63]: Minor and non-blocking: do we want to keep a deprecated alias for `CompositeFunctionCatalog`? It has been exported from the published package since 21.3.0, so removing it would be a source break for existing importers.

```ts
/** @deprecated Use `LocalFunctionCatalog`. */
export type CompositeFunctionCatalog = LocalFunctionCatalog;
```

The widened return types of `buildCompositeFunctionCatalogFromStepsAsync` and `buildLocalCompositeFunctionCatalogAsync` look like another possible source break: the loader still only returns `CompositeFunctionConfig`, but downstream code accessing `catalog[path].runs.steps` may stop compiling. Would it make sense to keep those return types composite-only until the loader starts returning legacy functions?

If we'd rather make both changes now, should they go under breaking changes in the changelog so the version bump comes out right?

**Review by hSATAC — COMMENTED** [review]
I took a pass through this and left two inline questions. I don't have much context on legacy custom build functions, so I'm mostly calling out a couple of compatibility differences I noticed rather than making a call on the intended behavior.

Neither is meant to block from my side, and I'll leave the final approval to Stanley since he has more context on the intended compatibility scope.

File: packages/eas-build-job/src/legacyFunction.ts (hunk 1 above)
Comment (sswrk) [reply packages/eas-build-job/src/legacyFunction.ts:57]: 👍 Good point, we should just keep full compatibility

File: packages/eas-build-job/src/localFunction.ts (hunk 2 above)
Comment (sswrk) [reply packages/eas-build-job/src/localFunction.ts:63]: Restored the old names via aliases to avoid breaking changes. My thinking was that we're fine to change this, thes packages are for internal stuff only, but right, this is public and can be used by some outside consumer.

File: packages/steps/src/CompositeFunctionExpander.ts [hunk 3]
```diff
@@ -66,7 +67,7 @@ type StepOverrides = {
 export class CompositeFunctionExpander {
   constructor(
     private readonly ctx: BuildStepGlobalContext,
-    private readonly compositeFunctionCatalog: CompositeFunctionCatalog,
+    private readonly compositeFunctionCatalog: LocalFunctionCatalog,
```
Comment (sjchmiela) [inline packages/steps/src/CompositeFunctionExpander.ts]:
```suggestion
    private readonly localFunctionCatalog: LocalFunctionCatalog,
```
?
